SwanCustomEnvironments/swancustomenvironments/scripts/makenv.sh

```diff
@@ -0,0 +1,164 @@
+    _log "  -h, --help                  Print this help page"
+    _log "  --accpy VERSION             Version of Acc-Py to be used"
+}
+
+# --------------------------------------------------------------------------------------------
+
+# Parse command line arguments
+while [ $# -gt 0 ]; do
+    key="$1"
+    case $key in
+        --env|-e)
+            ENV_NAME=$2
+            shift
+            shift
+            ;;
+        --repo|-r)
+            REPOSITORY=$2
+            shift
+            shift
+            ;;
+        --accpy)
+            ACCPY_VERSION=$2
+            shift
+            shift
+            ;;
+        --help|-h)
+            print_help
+            exit 0
+            ;;
+        *)
+            >&2 _log "ERROR: Invalid argument: $1" && _log
+            print_help
+            exit 1
+            ;;
+    esac
+done
+
+# --------------------------------------------------------------------------------------------
+
+# Checks if a name for the environment is given
+if [ -z "$ENV_NAME" ]; then
+    >&2 _log "ERROR: No virtual environment name provided." && _log
+    print_help
+    exit 1
+fi
+
+# Checks if the provided Acc-Py version is valid
+if [ -n "$ACCPY_VERSION" ] && [ ! -e "$ACCPY_PATH/base/$ACCPY_VERSION" ]; then
+    >&2 _log "ERROR: Invalid Acc-Py version."
+    exit 1
+fi
+
+# Checks if a repository file is given
+if [ -z "$REPOSITORY" ]; then
+    >&2 _log "ERROR: No repository provided." && _log
+    print_help
+    exit 1
+# Checks if the provided repository source is found
+elif [ -d $REPOSITORY ]; then
+    REQ_PATH="${REPOSITORY}/requirements.txt"
+    if [ ! -f "${REQ_PATH}" ]; then
+        >&2 _log "ERROR: Requirements file not found (${REQ_PATH})."
+        exit 1
+    fi
+elif [[ $REPOSITORY == http* ]]; then
+    # Extract the repository name from the URL
+    repo_name=$(basename $REPOSITORY)
+    repo_name=${repo_name%.*}
+    
+    mkdir -p $HOME/SWAN_projects
+
+    define_repo_path $repo_name
+
+    # Clone the repository
+    echo "Cloning the repository from ${REPOSITORY}..."
+    git clone $REPOSITORY -q "${REPO_PATH}" || { >&2 _log "ERROR: Failed to clone repository"; exit 1; }
+
+    REQ_PATH=${REPO_PATH}/requirements.txt
+    # Check if requirements.txt exists in the repository
+    if [ ! -f "${REQ_PATH}" ]; then
+        rm -rf ${REPO_PATH}
+        >&2 _log "ERROR: ${REQ_PATH} not found in ${REPO_PATH}."
+        exit 1
+    fi
+else
+    >&2 _log "ERROR: Invalid repository (${REPOSITORY})."
+    exit 1
+fi
+
+
+# --------------------------------------------------------------------------------------------
+
+ENV_PATH="/home/$USER/${ENV_NAME}"
+
+# Create virtual environment (acc-py or generic)
+if [ -n "$ACCPY_VERSION" ]; then
+    source $ACCPY_PATH/base/${ACCPY_VERSION}/setup.sh
+    if [ -d "${ENV_PATH}" ]; then
+        rm -rf ${ENV_PATH}
+    fi
+    acc-py venv ${ENV_PATH}
+else
+    message="Creating"
+    [ -d "${ENV_PATH}" ] && message="Recreating"
```

Comment:
We don't recreate anymore.

Reply:
Do we?

Reply:
> This starts to make even more sense with notebook URLs, since you can paste them several times and recreate multiple environments with differente configurations

But we shouldn't allow this, right? We must restrict to just one environment at a time in a SWAN session. And if a custom environment fails to start or the user goes back to the URL that triggers the CustomEnv extension, we ask the user if they would like to try again and if so, we delete their session and they can start another one again. Isn't this safer?

Reply:
Are you suggesting that when the user goes back `jupyter notebook -> customenvs extension`, the extension identifies if any session is created and, if so, pops up something like a modal saying something like `An environment created, do you want to reconfigure another one?` `Yes` (goes to spawner) `No` (goes back to where the user came from, most probably his lab default page).

Is this what you trying to say? Just to make it clear